libsplinter/src/circuit/template/rules/create_services.rs

```diff
@@ -50,3 +128,182 @@ impl From<v1::ServiceArgument> for ServiceArgument {
         }
     }
 }
+
+fn get_next_service_id(current_id: &str) -> Result<String, CircuitTemplateError> {
+    let alphanumeric = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
+    let mut next_id = current_id.to_string();
+    for (char_index, char) in current_id.char_indices().rev() {
+        let index = alphanumeric.find(char).ok_or_else(|| {
+            CircuitTemplateError::new(
+                "The field first_service must contain only valid base62 characters",
+            )
+        })?;
+        match alphanumeric.get(index + 1..index + 2) {
+            Some(sub_str) => {
+                let next_id_sub_string = next_id.get(char_index + 1..).unwrap_or_default();
+                let new_sub_string = format!("{}{}", sub_str, next_id_sub_string);
+                next_id.replace_range(char_index.., &new_sub_string);
+                return Ok(next_id);
+            }
+            None => {
+                let next_id_sub_string = next_id.get(char_index + 1..).unwrap_or_default();
+                let new_sub_string =
+                    format!("{}{}", alphanumeric[0..1].to_string(), next_id_sub_string);
+                next_id.replace_range(char_index.., &new_sub_string);
+            }
+        }
+    }
+
+    Err(CircuitTemplateError::new(
+        "Exceed number of services that can be built",
+    ))
+}
+
+fn all_services(
+    service_builders: Vec<SplinterServiceBuilder>,
+) -> Result<Vec<SplinterServiceBuilder>, CircuitTemplateError> {
+    let peers = service_builders.iter().try_fold::<_, _, Result<Vec<String>, CircuitTemplateError>>(Vec::new(), |mut acc, builder| {
+        let service_id = builder.service_id()
+            .ok_or_else(|| {
+                error!("The service_id must be set before the service argument PEER_SERVICES can be set");
+                CircuitTemplateError::new("Failed to parse template due to an internal error")
+            })?;
+        acc.push(service_id);
+        Ok(acc)
+    })?;
```

Comment:
Instead of using `.try_fold(...)?` here, you can use `.map(...).collect::<Result<Vec<String>, _>>()?` to simplify things a bit.